### PR TITLE
Implement support for the `REFTRAJ` simpulation

### PR DESCRIPTION
**Pull request by cpignedoli**
TrajectoryData added as spec.inputs to the cp2k plugin.
This is used for REFTRAJECTORY type of calculations.
The input trajectory will trigger teh creation of two input files:
-trajectory.xyz
-reftraj.cell

### CODE REVIEW

File: aiida_cp2k/calculations/__init__.py [hunk 1]
```diff
@@ -270,6 +286,16 @@ def prepare_for_submission(self, folder):
                 conflicting_keys=["COORDINATE"],
             )
 
+        # Craeate input trajectory files
```
Comment (yakutovicha) [inline aiida_cp2k/calculations/__init__.py:289]:
```suggestion
        # Create input trajectory files
```

File: aiida_cp2k/calculations/__init__.py [hunk 2]
```diff
@@ -270,6 +286,16 @@ def prepare_for_submission(self, folder):
...
+        if "trajectory" in self.inputs:
+            self._write_trajectories(
+                self.inputs.structure,
```
Comment (yakutovicha) [inline aiida_cp2k/calculations/__init__.py:292]: Why do you need the structure here? Isn't all the information available in the trajectory itself?

File: aiida_cp2k/calculations/__init__.py [hunk 3]
```diff
@@ -415,3 +451,29 @@ def _atoms_to_xyz(atoms):
...
+        xyz += f"{len(elem_coords)}\n"
+        xyz += f"i = {i+1} , time = {(i+1)*0.5} \n"
+        xyz += "\n".join(map(add, elem_symbols, elem_coords))
+        xyz += "\n"
+    if "cells" in trajectory.get_arraynames():
+        cell_vecs = [
+            f"{i+1} {(i+1)*0.5:6.3f} {p[0][0]:25.16f} {p[0][1]:25.16f} {p[0][2]:25.16f} {p[1][0]:25.16f} {p[1][1]:25.16f} {p[1][2]:25.16f} {p[2][0]:25.16f} {p[2][1]:25.16f} {p[2][2]:25.16f} {np.dot(p[0],np.cross(p[1],p[2]))}"
+            for (i, p) in enumerate(trajectory.get_array("cells"))
+        ]
+        cell += "\n".join(cell_vecs)
```
Comment (yakutovicha) [inline aiida_cp2k/calculations/__init__.py:478]: Rewrite this function. The `TrajectoryData` class has a method to get the structure from from a given frame.

File: aiida_cp2k/utils/input_generator.py [hunk 4]
```diff
@@ -207,9 +207,19 @@ def add_wfn_restart_section(input_dict, is_kpoints):
...
 @calcfunction
-def add_ext_restart_section(input_dict):
+def add_ext_restart_section(input_dict, first_snapshot=None):
     """Add external restart section to the input dictionary."""
     params = input_dict.get_dict()
+    if first_snapshot is not None:
+        params["MOTION"]["MD"]["REFTRAJ"]["FIRST_SNAPSHOT"] = first_snapshot
     # overwrite the complete EXT_RESTART section if present
     params["EXT_RESTART"] = {"RESTART_FILE_NAME": "./parent_calc/aiida-1.restart"}
+    if params["GLOBAL"]["RUN_TYPE"] == "MD":
```
Comment (yakutovicha) [inline aiida_cp2k/utils/input_generator.py:217]: Should we not just enable it always?

File: examples/single_calculations/example_dft_md_reftraj.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    }
+    builder.metadata.options.resources = {
+        "num_machines": 1,
+        "num_mpiprocs_per_machine": 1,
+    }
+    builder.metadata.options.max_wallclock_seconds = 1 * 3 * 60
+
+    print("Submitted calculation...")
+    run(builder)
+
```
Comment (yakutovicha) [inline examples/single_calculations/example_dft_md_reftraj.py:164]: Please add some quantities to compare with. It is a bit pointless to run a calculation without checking its results.

File: examples/single_calculations/example_dft_md_reftraj.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
+###############################################################################
```
Comment (yakutovicha) [inline examples/single_calculations/example_dft_md_reftraj.py:1]: Maybe this example is unnecessary as this run is also included in the `example_base_md_reftraj_restart.py` file.

File: aiida_cp2k/utils/input_generator.py (hunk 4 above)
Comment (cpignedoli) [reply aiida_cp2k/utils/input_generator.py:217]: what whould you like to enable always? FIRST_SNAPSHOT exist only in the REFTRAJ ensemble, it is not a general MD feature

File: aiida_cp2k/calculations/__init__.py (hunk 2 above)
Comment (cpignedoli) [reply aiida_cp2k/calculations/__init__.py:292]: no the trajectory does not carry information on the initial magnetization guess or, if needed in the future, ghost atoms

File: examples/single_calculations/example_dft_md_reftraj.py (hunk 6 above)
Comment (cpignedoli) [reply examples/single_calculations/example_dft_md_reftraj.py:1]: I agree I will remove it

File: aiida_cp2k/calculations/__init__.py [hunk 7]
```diff
@@ -59,6 +63,12 @@ def define(cls, spec):
             required=False,
             help="The main input structure.",
         )
+        spec.input(
+            "trajectory",
+            valid_type=TrajectoryData,
+            required=False,
+            help="Possible input trajectory fro REFTRAJMD.",
```
[the remark this answers is not in the document]
Comment (cpignedoli) [reply aiida_cp2k/calculations/__init__.py]: removed

File: aiida_cp2k/calculations/__init__.py (hunk 3 above)
Comment (yakutovicha) [reply aiida_cp2k/calculations/__init__.py]: I guess this is done in [06a33f3](https://github.com/aiidateam/aiida-cp2k/pull/207/commits/06a33f311c627228536f5e8963aa56da173eeabc).

File: examples/single_calculations/example_dft_md_reftraj.py (hunk 6 above)
Comment (cpignedoli) [reply examples/single_calculations/example_dft_md_reftraj.py]: done in [48f41b7](https://github.com/aiidateam/aiida-cp2k/pull/207/commits/48f41b7503c483ba2bc193274d451ec8a0311bb8)

File: aiida_cp2k/utils/input_generator.py (hunk 4 above)
Comment (yakutovicha) [reply aiida_cp2k/utils/input_generator.py]: For the moment, we agreed to keep them all set explicitly. If an issue pops up, we will revisit this section.

Especially, we need to monitor the replica chain calculations, which use constraints.

File: aiida_cp2k/calculations/__init__.py (hunk 2 above)
Comment (yakutovicha) [reply aiida_cp2k/calculations/__init__.py]: Despite @cpignedoli didn't like my idea, he removed this argument in [06a33f3](https://github.com/aiidateam/aiida-cp2k/pull/207/commits/06a33f311c627228536f5e8963aa56da173eeabc) 😆

File: examples/single_calculations/example_dft_md_reftraj.py (hunk 5 above)
Comment (yakutovicha) [reply examples/single_calculations/example_dft_md_reftraj.py]: Added a check for the total number of steps in all trajectories in [5ac5e24](https://github.com/aiidateam/aiida-cp2k/pull/207/commits/5ac5e24e90793790f67ddcf50771107493539a69)

**Review by yakutovicha — DISMISSED** [review]
Provided that the issue with the failed work chain must be fixed [upstream](https://github.com/aiidateam/aiida-core/issues/6307), we go ahead with a failed test for the moment.

**Review by yakutovicha — APPROVED** [review]

**Pull request merged**